### ISSUE TITLE
model parameters contains a lot of 'nan' at certain stage of training using AdamOptimizer

### ISSUE THREAD

**Issue by jiaboli007** [tensorflow/tensorflow#1812]
I am training a deep neural network using a simple and well defined case. The training goes very well and reach a high accuracy (~95%), and then, suddenly, just one step iteration, the accuracy is drop to almost zero. I then printed out all the model parameters, and there are lot of 'nan', something like this 
....0.076455489, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 0.08646702....

Here is the information about my system.

1) Installation. I installed tensor flow from source code on AWS ubuntu g2 instance (GPU instance). During the installation, I use customized configuration so that the tensor flow can run on g2 instance with computing capacity 3.0. 

2) DNN structure: I have a 30 hidden fully connected layers. Each layer has 32 nodes. To make the DNN more stable, I have added short-cuts for every three layers (similar to a recent publication from Microsoft Research "Deep Residual Learning for Image Recognition", by Kaiming He et al.) 

3) Optimizer: AdamOptimizer. 
  I first found the problem with AdamOptimzer. However, I just realized that the same problem also happens to GradientDescentOptimizer.

4) Procedure to produce the bug. 

I have a saved model, which has good accuracy, and I restart the training by reading the saved model, and the problem shows up for the first iteration. 

I wonder if this is a known issue or not. I would be happy to provide my code and input data if this is necessary. I need to create a simple case with simplified code as I am working on a production code that I cannot post it here.

**Comment by johnrowlay** [comment]
I'm not hugely experienced with either neural networks or Tensor Flow, but it sounds like it could be the "vanishing gradients" problem. When you have a very deep network (and30 layers is *very* deep), the gradients in the loss function become very small, and I have seen elsewhere that this can cause NaN values. It sounds like in the single step you mention, the gradients reduce and cross a precision threshold, and therefore become undefined.

**Comment by johnrowlay** [comment]
As for a solution, you should just try decreasing the number of layers in your network. (Possibly) increasing the precision of your data might help, e.g. to ft.float64.

**Comment by martinwicke** [comment]
This type of question is better handled at StackOverflow, please ask it there. We're trying to keep github issues for bug reports. Thanks.

**Comment by jiaboli007** [comment]
I report it as a bug. I will post a simple code for reproducing the bug.

**Comment by martinwicke** [comment]
Yes, please post some code to reproduce. Getting NaNs during training is quite common and not usually a sign of a bug. A small code snippets that shows an actual bug will help a lot. I'll reopen this once you post code to reproduce.

**Comment by jiaboli007** [comment]
Hi Martin,

Here is the code for reproducing the bug. I have made the code much simpler, and only have two hidden layers, a very moderate NN. 

To run the code, type

python TFDebug.py

After about 5000 iterations, the bad model parameters will be printed out. 

I have tested it on all platforms that I have (ubuntu CPU, ubuntu GPU and Mac OS).

All show the same type of error. 

Many thanks.

BWT, since I have the source code, if you can point out which codes are related to the problem, I can also try to debug it.

Jiabo

[TFDebug.txt](https://github.com/tensorflow/tensorflow/files/209258/TFDebug.txt)

**Comment by martinwicke** [comment]
Please use the check_numerics op to find out where the NaNs are first introduced. In particular, split the .optimize() call on the optizer into compute_gradients and apply_gradients, so you can look at the computed gradients. If you see any variable get really large or very small, that's a sign of trouble. 

The distance function you wrote may be the cause, you could try making sure the value inside the square root is greater then 0 before you take the square root.

**Comment by yaroslavvb** [comment]
I think it's a bug in your code rather than in TensorFlow. Your confidences converge to 0's and 1's, then you take a log of 0 which gives -Inf, and multiply it by 0, which gives NaN

**Comment by martinwicke** [comment]
Thanks! I got too hypnotized by the sqrt to see the log. :)

**Comment by jiaboli007** [comment]
Thanks, I will take a closer look at the code.

Yaroslavb, could you explain your idea in a little more detail? I am not sure what you mean "confidence converge to 0's and 1's.

**Comment by yaroslavvb** [comment]
You are taking a tf.log(y), and you get a 0 as an element of y as you train

**Comment by jiaboli007** [comment]
Good point. Thanks.  However, y comes from a softmax function, which should always return a positive value, right?  I will check if some elements in y is actually rounded to zeros. Any simple trick to avoid log(0)? Add a very small positive number?

I also check the distance function, and I don't think the problem comes from it. To make sure that is not a problem, I even put a check before taking sqrt. 

I will dig a little more deeper into the code, following Martin's suggestion.

**Comment by yaroslavvb** [comment]
Mathematically, softmax can never be zero, but because we use floating point numbers, a number close enough to zero is represented as zero. Anyway, this is a common issue with neural networks, search for NaN and softmax

**Comment by jiaboli007** [comment]
Many thanks. I appreciate very much your comments!

**Comment by jiaboli007** [comment]
I have verified that the problem is indeed from log(y). I added a small constant to y, and the problem is solved. Thanks to all.
